src/main/java/ch/njol/skript/expressions/ExprPermissions.java

```diff
@@ -31,58 +37,73 @@
-@Description("Returns all permissions of the defined player(s). Note that the modifications to resulting list do not actually change permissions.")
+@Description("Returns all permissions of the defined permissible(s). A permissible is an object like an entity that can have permissions.")
 @Examples("set {_permissions::*} to all permissions of the player")
-@Since("2.2-dev33")
+@Since("2.2-dev33, INSERT VERSION (Changers)")
 public class ExprPermissions extends SimpleExpression<String> {
-	
+
 	static {
-		Skript.registerExpression(ExprPermissions.class, String.class, ExpressionType.PROPERTY, "[(all [[of] the]|the)] permissions (from|of) %players%", "[(all [[of] the]|the)] %players%'[s] permissions");
+		Skript.registerExpression(ExprPermissions.class, String.class, ExpressionType.PROPERTY, "[(all [[of] the]|the)] permissions (from|of) %entities%", "[(all [[of] the]|the)] %entities%'[s] permissions");
 	}
-	
-	@SuppressWarnings("null")
-	private Expression<Player> players;
-	
-	@SuppressWarnings({"null", "unchecked"})
+
+	// Metadata tag
+	private static final String PERMISSION_TAG = "skript-permissions";
+
+	private Expression<Entity> entities;
+
 	@Override
-	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
-		players = (Expression<Player>) exprs[0];
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		entities = (Expression<Entity>) exprs[0];
 		return true;
 	}
 
 	@Override
 	@Nullable
-	protected String[] get(Event e) {
-		final Set<String> permissions = new HashSet<>();
-		for (Player player : players.getArray(e))
-			for (final PermissionAttachmentInfo permission : player.getEffectivePermissions())
-				permissions.add(permission.getPermission());
-		return permissions.toArray(new String[permissions.size()]);
+	protected String[] get(Event event) {
+		return entities.stream(event)
+				.flatMap(permissible -> permissible.getEffectivePermissions().stream())
+				.map(permission -> permission.getPermission())
+				.toArray(String[]::new);
 	}
-	
+
 	@Override
 	public boolean isSingle() {
 		return false;
 	}
-	
+
 	@Override
 	public Class<String> getReturnType() {
 		return String.class;
 	}
-	
+
 	@Override
 	public String toString(@Nullable Event event, boolean debug) {
-		return "permissions " + " of " + players.toString(event, debug);
+		return "permissions " + " of " + entities.toString(event, debug);
+	}
+
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		if (mode == ChangeMode.ADD || mode == ChangeMode.REMOVE)
+			return CollectionUtils.array(String.class, String[].class);
+		return null;
+	}
+
+	@Override
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		for (Entity entity : entities.getAll(event)) {
+			PermissionAttachment perm = getPermission(entity);
+			for (String string : (String[]) delta)
```

Comment:
Casting delta itself can cause issues sometimes (it has happened with skript-gui before)

Reply:
I don't see it here.

Reply:
It only happens when you spam the expression. This seems like an issue would need to address
```
java.lang.ClassCastException: class [Ljava.lang.Object; cannot be cast to class [Ljava.lang.String; ([Ljava.lang.Object; and [Ljava.lang.String; are in module java.base of loader 'bootstrap')
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.expressions.ExprPermissions.change(ExprPermissions.java:108)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.effects.EffChange.execute(EffChange.java:280)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.lang.Effect.run(Effect.java:50)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.lang.TriggerItem.walk(TriggerItem.java:62)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.lang.TriggerItem.walk(TriggerItem.java:90)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.sections.EffSecSpawn.lambda$walk$1(EffSecSpawn.java:149)
[21:32:12 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R1.CraftRegionAccessor.addEntity(CraftRegionAccessor.java:533)
[21:32:12 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R1.CraftRegionAccessor.spawn(CraftRegionAccessor.java:512)
[21:32:12 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R1.CraftRegionAccessor.spawn(CraftRegionAccessor.java:506)
[21:32:12 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R1.CraftRegionAccessor.spawn(CraftRegionAccessor.java:497)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.entity.EntityData.spawn(EntityData.java:426)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.sections.EffSecSpawn.walk(EffSecSpawn.java:167)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.lang.TriggerItem.walk(TriggerItem.java:90)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.lang.Trigger.execute(Trigger.java:52)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.events.EvtScript.runTrigger(EvtScript.java:88)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.events.EvtScript.register(EvtScript.java:68)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.lang.SkriptEvent.postLoad(SkriptEvent.java:151)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.ScriptLoader.lambda$loadScripts$9(ScriptLoader.java:573)
[21:32:12 ERROR]: #!#!     at java.base/java.util.ArrayList.removeIf(ArrayList.java:1672)
[21:32:12 ERROR]: #!#!     at java.base/java.util.ArrayList.removeIf(ArrayList.java:1660)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.ScriptLoader.lambda$loadScripts$10(ScriptLoader.java:569)
[21:32:12 ERROR]: #!#!     at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:684)
[21:32:12 ERROR]: #!#!     at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:662)
[21:32:12 ERROR]: #!#!     at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2168)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.ScriptLoader.loadScripts(ScriptLoader.java:513)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.ScriptLoader.loadScripts(ScriptLoader.java:460)
[21:32:12 ERROR]: #!#!     at Skript.jar//ch.njol.skript.SkriptCommand.onCommand(SkriptCommand.java:189)
[21:32:12 ERROR]: #!#!     at org.bukkit.command.PluginCommand.execute(PluginCommand.java:45)
[21:32:12 ERROR]: #!#!     at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:155)
[21:32:12 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R1.CraftServer.dispatchCommand(CraftServer.java:916)
[21:32:12 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R1.CraftServer.dispatchServerCommand(CraftServer.java:879)
[21:32:12 ERROR]: #!#!     at net.minecraft.server.dedicated.DedicatedServer.bg(DedicatedServer.java:500)
[21:32:12 ERROR]: #!#!     at net.minecraft.server.dedicated.DedicatedServer.b(DedicatedServer.java:447)
[21:32:12 ERROR]: #!#!     at net.minecraft.server.MinecraftServer.a(MinecraftServer.java:1397)
[21:32:12 ERROR]: #!#!     at net.minecraft.server.MinecraftServer.v(MinecraftServer.java:1173)
[21:32:12 ERROR]: #!#!     at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:305)
[21:32:12 ERROR]: #!#!     at java.base/java.lang.Thread.run(Thread.java:833)
```